### PR TITLE
Initial upgrade procedure documentation

### PR DESCRIPTION
**Pull request by Lukasa**
This is my first draft of documentation for our upgrade procedure.

### CODE REVIEW

File: docs/source/opens-upgrade.rst [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+
+Ubuntu 14.04
+^^^^^^^^^^^^
+
+First, upgrade packaged components::
+
+    apt-get update
+    apt-get install dnsmasq-base nova-api-metadata neutron-dhcp-agent neutron-common python-etcd calico-compute
+
+Finally, restart Felix to ensure that it picks up any changes::
```
Comment (alexwlchan) [inline docs/source/opens-upgrade.rst]: Nitpick:having both "first ... finally" feels weird to me. "finally" sounds as if we're at the end of a much longer list; I'd just drop it. "Then restart Felix..."

File: docs/source/opens-upgrade.rst [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+^^^^^^^^^^^^
+
+First, update packaged components::
+
+    apt-get update
+    apt-get install python-etcd calico-control
+
+Finally, restart Neutron to ensure that it picks up any changes::
+
+    service neutron-server restart
```
Comment (alexwlchan) [inline docs/source/opens-upgrade.rst]: Nitpick: ditto above.

File: docs/source/opens-upgrade.rst [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+but will likely be delayed until the end of the upgrade.
+
+For this reason, we highly recommend planning a maintenance window for the
+upgrade. During this window, you should disable all user API access to your
+OpenStack deployment.
+
+Upgrade Procedure
+-----------------
+
+Upgrade is performed in the following stages.
```
Comment (ed-harrison) [inline docs/source/opens-upgrade.rst]: "stages, which must be performed in the order shown."  [Worth hammering home this point, I think.]

File: docs/source/opens-upgrade.rst [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+1: Upgrade etcd
+~~~~~~~~~~~~~~~
+
+This step should be run on every machine in your deployment that runs any
+Calico code, and also on the machine running the etcd cluster.
+
+Ubuntu 14.04
+^^^^^^^^^^^^
+
+On the machine running etcd, simply use apt-get to obtain the more recent
```
Comment (ed-harrison) [inline docs/source/opens-upgrade.rst]: Personally, I'd remove "simply" here.

File: docs/source/opens-upgrade.rst [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+    apt-get install etcd
+
+Red Hat 7
+^^^^^^^^^
+
+On the machine running etcd, begin by stopping the etcd process::
+
+    systemctl stop etcd
+
+Then, download the latest binary and install it::
```
Comment (ed-harrison) [inline docs/source/opens-upgrade.rst]: Here you say download the "latest" binary, but then give the instruction to download 2.0.11.  If we want to pin ourselves to 2.0.11, then say that in the line above (e.g. "Download the correct etcd binary - Calico is currently tested to work with etcd v2.0.11.").

File: docs/source/opens-upgrade.rst [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+
+First, upgrade python-etcd::
+
+    curl -L https://github.com/Metaswitch/python-etcd/archive/master.tar.gz -o python-etcd.tar.gz
+    tar xvf python-etcd.tar.gz
+    cd python-etcd-master
+    python setup.py install
+
+Then, update the relevant components::
+
```
Comment (ed-harrison) [inline docs/source/opens-upgrade.rst]: Will yum update update other things beyond Calico?  Just thinking that in the Ubuntu instructions we're explicitly upgrading only the Calico stuff, but if yum update could upgrade other packages, that could have unintended consequences.  Is there a way to just update the Calico stuff?

File: docs/source/opens-upgrade.rst [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+    yum update
```
Comment (ed-harrison) [inline docs/source/opens-upgrade.rst]: ditto above

**Comment by ed-harrison** [comment]
I think you should have a pointer to the upgrade instructions at the top of the relevant install instructions.  e.g. "These instructions will take you through first time installation of Calico...  If you want to upgrade an existing version of Calico, then see..."

Do you think we need to say anything about the upgrade instructions only working to upgrade versions after 0.xx (whatever xx is, in practice)?

File: docs/source/opens-upgrade.rst (hunk 6 above)
Comment (alexwlchan) [inline docs/source/opens-upgrade.rst]: @ed-harrison Yes and yes. From the [Red Hat docs](https://access.redhat.com/documentation/en-US/Red_Hat_Network_Satellite/5.5/html/Reference_Guide/sect-Reference_Guide-Package_Updater-Updating_Packages_from_the_Command_Line_with_yum.html):

> If no packages are specified, then yum will attempt to update all installed packages.

Fix is to do <code>yum update <em>calico-package-name</em></code> for appropriate package names.

**Comment by Lukasa** [comment]
Thanks @ed-harrison @alexwlchan, those markups have been made.

@ed-harrison, I don't think we need that tag in the upgrade instructions at this time. I think we consider the pre-etcd days to be the dark past of this project, and so we just don't really want to worry about it.

**Comment by ed-harrison** [comment]
LGTM

**Pull request merged**